code/game/objects/items/RSF.dm

```diff
@@ -8,178 +8,201 @@ RSF
+		var/tempMatter = matter_by_item[W.type] + matter
+		if(tempMatter > max_matter)
+			to_chat(user, "<span class='warning'>\The [src] can't hold any more [discriptor]!</span>")
 			return
 		qdel(W)
-		matter += 10
+		matter = tempMatter //We add its value
 		playsound(src.loc, 'sound/machines/click.ogg', 10, TRUE)
-		to_chat(user, "<span class='notice'>The RSF now holds [matter]/30 fabrication-units.</span>")
-		icon_state = "rsf"
+		to_chat(user, "<span class='notice'>\The [src] now holds [matter]/[max_matter] [discriptor].</span>")
+		icon_state = initial(icon_state)//and set the icon state to the initial value it had
 	else
 		return ..()
 
 /obj/item/rsf/attack_self(mob/user)
+	if(!user)
+		return
 	playsound(src.loc, 'sound/effects/pop.ogg', 50, FALSE)
-	switch(mode)
-		if(5)
-			mode = 1
-			to_chat(user, "<span class='notice'>Changed dispensing mode to 'Drinking Glass'.</span>")
-		if(1)
-			mode = 2
-			to_chat(user, "<span class='notice'>Changed dispensing mode to 'Paper'.</span>")
-		if(2)
-			mode = 3
-			to_chat(user, "<span class='notice'>Changed dispensing mode to 'Pen'.</span>")
-		if(3)
-			mode = 4
-			to_chat(user, "<span class='notice'>Changed dispensing mode to 'Dice Pack'.</span>")
-		if(4)
-			mode = 5
-			to_chat(user, "<span class='notice'>Changed dispensing mode to 'Cigarette'.</span>")
+	var/atom/target = cost_by_item
+	var/cost = 0
+	//Warning, prepare for bodgecode
+	while(istype(target, /list/))//While target is a list we continue the loop
+		var/picked = show_radial_menu(user, src, formRadial(target), custom_check = CALLBACK(src, .proc/check_menu, user), require_near = TRUE)
+		if(!check_menu(user) || picked == null)
+			return
+		for(var/emem in target)//Back through target agian
+			var/atom/test = extractObject(target, emem)
+			if(picked == test.name)//We try and find the entry that matches the radial selection
+				cost = target[emem]//We cash the cost
+				target = emem
+				break
+			qdel(test)
+		//If we found a list we start it all again, this time looking through its contents.
+	to_dispense = target
+	dispense_cost = cost
 	// Change mode
 
+///Extracts the related object from a associated list of objects and values, or lists and objects.
+/obj/item/rsf/proc/extractObject(from, input)
+	var/atom/test
+	if(istype(input, /list/))
+		var/temp = from[input]//If it's a list we should use its associated object
+		test = new temp()
+	else
+		test = new input()
+	return test
+
+///Forms a radial menu based off an object in a list, or a list's associated object
+/obj/item/rsf/proc/formRadial(from)
+	var/list/radial_list = list()
+	for(var/meme in from)//We iterate through all of targets entrys
+		var/atom/test = extractObject(from, meme)
+		//We then add their data into the radial menu
+		radial_list[test.name] = image(icon = test.icon, icon_state = test.icon_state)
+		qdel(test)
+	return radial_list
+
+/obj/item/rsf/proc/check_menu(mob/user)
+	if(!istype(user))
+		return FALSE
```

Comment:
Can this realistically happen? Make it `CRASH` if so.

Reply:
See the other discussion about CRASH. It's a holdover from the RCD radial code.